crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
+            for position in MaspDigitPos::iter() {
+                let native_asset_epoch0 = AssetData {
+                    token: native_token.clone(),
+                    denom: native_token_denom,
+                    position,
+                    epoch: Some(epoch_zero),
+                };
+                let native_asset_epoch0_type = native_asset_epoch0
+                    .encode()
+                    .map_err(|_| eyre!("unable to create asset type"))?;
+                let native_asset = AssetData {
+                    token: native_token.clone(),
+                    denom: native_token_denom,
+                    position,
+                    epoch: Some(epoch),
+                };
+                let native_asset_type = native_asset
+                    .encode()
+                    .map_err(|_| eyre!("unable to create asset type"))?;
+
+                // Pre-date all the assets to MaspEpoch(0) for comparison
+                if let Some((_, amt)) = current_exchanged_balance
+                    .project(native_asset_type)
+                    .into_components()
+                    .last()
+                {
+                    current_native_balance +=
+                        ValueSum::from_pair(native_asset_epoch0_type, amt);
+                }
+                if let Some((_, amt)) = next_exchanged_balance
+                    .project(native_asset_type)
+                    .into_components()
+                    .last()
+                {
+                    next_native_balance +=
+                        ValueSum::from_pair(native_asset_epoch0_type, amt);
+                }
+            }
         }
+
+        let rewards = next_native_balance - current_native_balance;
+        let decoded_rewards =
+            self.decode_sum(context.client(), rewards).await.0;
+
+        decoded_rewards
+            .into_components()
+            .try_fold(
+                Amount::zero(),
+                |acc,
+
+                 (
+                    (
+                        _,
+                        AssetData {
+                            token,
+                            denom: _,
+                            position,
+                            epoch,
+                        },
+                    ),
+                    amt,
+                )| {
+                    // Sanity checks, the rewards must be given in the native
+                    // asset at masp epoch 0
+                    if token != native_token {
+                        return Err(eyre!(
+                            "Found reward asset other than the native token"
+                        ));
+                    }
+                    match epoch {
+                        Some(ep) if ep == epoch_zero => (),
+                        _ => {
+                            return Err(eyre!(
+                                "Found reward asset with an epoch different \
+                                 than the current one"
```

Comment:
This error message doesn't seem correct.

Reply:
I feel like you left me hanging here. You checked if the epoch is zero and error otherwise.

Reply:
So we extract the native token components from the result of the subtraction of the two balances. Then, to be able to sum all these components and return  a `DenominatedAmount`, we normalize them at `MaspEpoch(0)`. But we could pick any other epoch so long as it's a fixed one. The new error message is:

`Found reward asset with an epoch different than the 0th`

Should I change it to something else?

Reply:
I'm good with that